### ISSUE TITLE
Bert attention Mask 관련 정리

### ISSUE THREAD

**Issue by suuperstone** [suuperstone/issue#4]

**Comment by suuperstone** [comment]
## Bert Encoder

attention mask 를 attention_scores 에 직접 더해지는 형태로 줘야함.
https://huggingface.co/transformers/v3.5.1/_modules/transformers/modeling_bert.html

<img width="1116" alt="image" src="https://user-images.githubusercontent.com/59866074/186842032-a5e2c13f-a6cd-4422-b3cf-fe2095051325.png">
<img width="1122" alt="image" src="https://user-images.githubusercontent.com/59866074/186842074-2a472f46-5e00-48ef-9504-2d95786d66fd.png">

attention_score 의 형태가 ( B, num_heads, max_length, max_length ) 
형태이기 때문에 ( B, 1, max_length, max_length ) 형태로 mask 를 줘야함.
-> first dimension 을 num_heads 로 확장하면 head 별로 masking 이 가능할듯.

## Bert Model

attention_mask 가 _get_extended_attention_mask 의 함수를 처리하고 나온다
output 의 형태는  ( B, 1, 1, max_length) 가 된다.

https://github.com/huggingface/transformers/blob/06a6a4bd516f7d0ba7c4966a2d3d9c0bf07797ae/src/transformers/models/bert/modeling_bert.py#L995

https://github.com/huggingface/transformers/blob/06a6a4bd516f7d0ba7c4966a2d3d9c0bf07797ae/src/transformers/modeling_utils.py#L700

<img width="1007" alt="image" src="https://user-images.githubusercontent.com/59866074/186842542-8497c30a-0b32-44c8-ae73-307d85e8b820.png">

위의 경우와 다른 attention_mask 값이 들어감.

그래서 이 경우 attention mask 가 씌워지는 형태가 우리 생각하고 조금 다르게 씌워지고 있다는 걸 알 수 있음.

(B, num_heads, max_length, max_length) + (B, 1, 1, max_length)
의 연산을 하면, 각 num_heads, 각 row (third dimension) 마다 (0,0,0,0..,.-inf, -inf,..) 를 더하는 형태가 된다.
그래서 사실 [1,1,1,1,1,0,0,0] 과 같은 mask 를 주게 되면
앞 쪽 다섯 개의 value 는 영향없이 정상적인 mask 가 주어진 값이 나오지만,
뒤쪽 3개의 embedding 은 정상적인 mask 가 없는 형태의 value 가 나오게된다. ( key masking 이 안됨. )
